lint: add append-combine checker

Detects `append` chains to the same slice that can be done in a single `append` call.

**Before:**
```go
xs = append(xs, 1)
xs = append(xs, 2)
```
**After:**
```go
xs = append(xs, 1, 2)
```

There are both performance and style reasons for choosing suggested form.
Even if compiler will rewrite "before" info "after", second form usually is easier to read with less
code repetition (especially if slice name is long).

Does trigger several times inside GOROOT:
```

$GOROOT/src/cmd/dist/test.go:605:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/dist/test.go:674:4: append-combine: found appends with 2 args
$GOROOT/src/cmd/cgo/util.go:48:3: append-combine: found appends with 3 args
$GOROOT/src/cmd/vet/asmdecl.go:425:4: append-combine: found appends with 2 args
$GOROOT/src/cmd/vet/asmdecl.go:430:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/vet/asmdecl.go:434:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/vet/asmdecl.go:438:3: append-combine: found appends with 3 args
$GOROOT/src/cmd/vet/asmdecl.go:443:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/vet/asmdecl.go:448:3: append-combine: found appends with 2 args
$GOROOT/src/crypto/tls/handshake_server_test.go:515:2: append-combine: found appends with 2 args
$GOROOT/src/cmd/link/internal/ld/dwarf.go:1397:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/link/internal/ld/lib.go:1208:2: append-combine: found appends with 2 args
$GOROOT/src/cmd/link/internal/ld/lib.go:1164:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/link/internal/ld/lib.go:1312:3: append-combine: found appends with 5 args
$GOROOT/src/crypto/x509/name_constraints_test.go:1728:2: append-combine: found appends with 2 args
$GOROOT/src/crypto/x509/name_constraints_test.go:1711:3: append-combine: found appends with 6 args
$GOROOT/src/strconv/quote.go:31:4: append-combine: found appends with 2 args
$GOROOT/src/strconv/quote.go:54:3: append-combine: found appends with 2 args
$GOROOT/src/strconv/quote.go:87:4: append-combine: found appends with 2 args
$GOROOT/src/net/http/socks_bundle.go:106:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/nm/nm_test.go:223:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/nm/nm_test.go:226:4: append-combine: found appends with 2 args
$GOROOT/src/cmd/nm/nm_test.go:229:4: append-combine: found appends with 2 args
$GOROOT/src/cmd/compile/internal/ssa/deadcode_test.go:146:4: append-combine: found appends with 2 args
$GOROOT/src/cmd/go/internal/envcmd/env.go:93:2: append-combine: found appends with 2 args
$GOROOT/src/archive/tar/writer_test.go:36:4: append-combine: found appends with 2 args
$GOROOT/src/cmd/compile/internal/gc/range.go:225:3: append-combine: found appends with 2 args
$GOROOT/src/cmd/compile/internal/gc/select.go:338:2: append-combine: found appends with 3 args
$GOROOT/src/cmd/compile/internal/gc/walk.go:3325:2: append-combine: found appends with 2 args
$GOROOT/src/cmd/compile/internal/gc/inl_test.go:152:3: append-combine: found appends with 3 args
$GOROOT/src/net/dnsclient.go:30:3: append-combine: found appends with 4 args
$GOROOT/src/net/mac.go:21:3: append-combine: found appends with 2 args
$GOROOT/src/net/interface_linux_test.go:19:2: append-combine: found appends with 2 args
$GOROOT/src/net/interface_linux_test.go:44:2: append-combine: found appends with 2 args
```

--- a/cmd/gocritic/testdata/append-combine/negative_tests.go
+++ b/cmd/gocritic/testdata/append-combine/negative_tests.go
@@ -1,0 +1,39 @@
+package checker_test
+
+func noWarnings1() {
+	var xs []int
+	var ys []int
+
+	// OK: can't combine if there is "..." argument.
+	xs = append(xs, ys...)
+	xs = append(xs, 1)
+
+	_ = 0
+
+	// OK: appends to different slices.
+	xs = append(xs, 1)
+	ys = append(ys, 2)
+}
+
+func noWarnings2() {
+	xs := map[string][]int{}
+
+	// OK: different keys.
+	xs["k1"] = append(xs["k1"], 1)
+	xs["k2"] = append(xs["k2"], 2)
+}
+
+func noWarnings3() {
+	var xs []int
+
+	// OK: different blocks.
+	xs = append(xs, 1)
+	{
+		xs = append(xs, 2)
+		{
+			xs = append(xs, 3)
+		}
+		// OK: chain interrupted by block above.
+		xs = append(xs, 4)
+	}
+}

--- a/cmd/gocritic/testdata/append-combine/positive_tests.go
+++ b/cmd/gocritic/testdata/append-combine/positive_tests.go
@@ -1,0 +1,72 @@
+package checker_test
+
+func warnings1() {
+	var xs []int
+
+	/// can combine chain of 2 appends into one
+	xs = append(xs, 1)
+	xs = append(xs, 2)
+
+	_ = 0
+
+	/// can combine chain of 2 appends into one
+	xs = append(xs, 1, 2)
+	xs = append(xs, 3, 4)
+
+	_ = 0
+
+	/// can combine chain of 3 appends into one
+	xs = append(xs, 1)
+	xs = append(xs, 2)
+	xs = append(xs, 3)
+
+	switch len(xs) == 0 {
+	case true:
+		/// can combine chain of 2 appends into one
+		xs = append(xs, 1)
+		xs = append(xs, 2)
+	case false:
+		/// can combine chain of 4 appends into one
+		xs = append(xs, 1)
+		xs = append(xs, 2)
+		xs = append(xs, 3)
+		xs = append(xs, 4, 5, 6)
+	default:
+		// Intermixing chains and breaks.
+
+		var ys []int
+		xs = append(xs, ys...)
+		/// can combine chain of 2 appends into one
+		xs = append(xs, 1, 2)
+		xs = append(xs, 3)
+		xs = append(xs, ys...)
+		xs = append(xs, 4)
+		xs = append(xs, ys...)
+		/// can combine chain of 3 appends into one
+		xs = append(xs, 5, 6)
+		xs = append(xs, 7, 8)
+		xs = append(xs, 9)
+	}
+
+	/// can combine chain of 3 appends into one
+	xs = append(xs, 1)
+	// Comments can't break the chain.
+	xs = append(xs, 2)
+	// Even if there are multiple.
+	xs = append(xs, 3)
+}
+
+func warnings2() {
+	xs := map[string][]int{}
+
+	/// can combine chain of 2 appends into one
+	xs["k"] = append(xs["k"], 1)
+	xs["k"] = append(xs["k"], 2)
+
+	xs["k1"] = append(xs["k1"], 1)
+	/// can combine chain of 2 appends into one
+	xs["k2"] = append(xs["k2"], 2)
+	xs["k2"] = append(xs["k2"], 3)
+	xs["k3"] = append(xs["k3"], 4)
+	xs["k2"] = append(xs["k2"], 5)
+}

--- a/cmd/gocritic/testdata/append-combine/positive_tests.go
+++ b/cmd/gocritic/testdata/append-combine/positive_tests.go
@@ -48,6 +48,27 @@ func warnings1() {
 		xs = append(xs, 9)
 	}
 
+	ch := make(chan bool)
+	select {
+	case <-ch:
+		/// can combine chain of 2 appends into one
+		xs = append(xs, 1)
+		xs = append(xs, 2)
+		if ch != nil {
+			/// can combine chain of 2 appends into one
+			xs = append(xs, 5)
+			xs = append(xs, 6)
+		} else {
+			/// can combine chain of 2 appends into one
+			xs = append(xs, 7)
+			xs = append(xs, 8)
+		}
+	default:
+		/// can combine chain of 2 appends into one
+		xs = append(xs, 3)
+		xs = append(xs, 4)
+	}
+
 	/// can combine chain of 3 appends into one
 	xs = append(xs, 1)
 	// Comments can't break the chain.

--- a/docs/checkers/append-combine.md
+++ b/docs/checkers/append-combine.md
@@ -1,0 +1,12 @@
+Detects `append` chains to the same slice that can be done in a single `append` call.
+
+**Before:**
+```go
+xs = append(xs, 1)
+xs = append(xs, 2)
+```
+
+**After:**
+```go
+xs = append(xs, 1, 2)
+```

--- a/docs/overview.md
+++ b/docs/overview.md
@@ -12,6 +12,10 @@ Go source code linter that brings checks that are currently not implemented in o
     <th>Experimental</th>
   </tr>
   <tr>
+    <td><a href="#append-combine-ref">append-combine</a></td>
+    <td>false</td>
+  </tr>
+  <tr>
     <td><a href="#builtin-shadow-ref">builtin-shadow</a></td>
     <td>false</td>
   </tr>
@@ -84,6 +88,22 @@ Go source code linter that brings checks that are currently not implemented in o
     <td>false</td>
   </tr>
 </table>
+
+
+<a name="append-combine-ref"></a>
+## append-combine
+Detects `append` chains to the same slice that can be done in a single `append` call.
+
+**Before:**
+```go
+xs = append(xs, 1)
+xs = append(xs, 2)
+```
+
+**After:**
+```go
+xs = append(xs, 1, 2)
+```
 
 
 <a name="builtin-shadow-ref"></a>

--- a/lint/append-combine_checker.go
+++ b/lint/append-combine_checker.go
@@ -1,0 +1,93 @@
+package lint
+
+import (
+	"go/ast"
+	"go/token"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+func appendCombineCheck(ctx *context) func(*ast.File) {
+	return wrapBlockChecker(&appendCombineChecker{
+		baseStmtListChecker: baseStmtListChecker{ctx: ctx},
+	})
+}
+
+type appendCombineChecker struct {
+	baseStmtListChecker
+}
+
+func (c *appendCombineChecker) CheckStmtList(list []ast.Stmt) {
+	var cause ast.Node // First append
+	var slice ast.Expr // Slice being appended to
+	chain := 0         // How much appends in a row we've seen
+
+	// Break the chain.
+	// If enough appends are in chain, print warning.
+	flush := func() {
+		if chain > 1 {
+			c.warn(cause, chain)
+		}
+		chain = 0
+		slice = nil
+	}
+
+	for _, stmt := range list {
+		call := c.matchAppend(stmt, slice)
+		if call == nil {
+			flush()
+			continue
+		}
+
+		if chain == 0 {
+			// First append in a chain.
+			chain = 1
+			slice = call.Args[0]
+			cause = stmt
+		} else {
+			chain++
+		}
+	}
+
+	// Required for printing chains that consist of trailing
+	// statements from the list.
+	flush()
+}
+
+func (c *appendCombineChecker) matchAppend(stmt ast.Stmt, slice ast.Expr) *ast.CallExpr {
+	// Seeking for:
+	//	slice = append(slice, xs...)
+	// xs are 0-N append arguments, but not variadic argument,
+	// because it makes append combining impossible.
+
+	assign, ok := stmt.(*ast.AssignStmt)
+	{
+		cond := ok &&
+			len(assign.Lhs) == 1 &&
+			len(assign.Rhs) == 1
+		if !cond {
+			return nil
+		}
+	}
+	call, ok := assign.Rhs[0].(*ast.CallExpr)
+	{
+		cond := ok &&
+			qualifiedName(call.Fun) == "append" &&
+			call.Ellipsis == token.NoPos &&
+			astcmp.EqualExpr(assign.Lhs[0], call.Args[0])
+		if !cond {
+			return nil
+		}
+	}
+
+	// Check that current append slice match previous append slice.
+	// Otherwise we should break the chain.
+	if slice == nil || astcmp.EqualExpr(slice, call.Args[0]) {
+		return call
+	}
+	return nil
+}
+
+func (c *appendCombineChecker) warn(cause ast.Node, chain int) {
+	c.ctx.Warn(cause, "can combine chain of %d appends into one", chain)
+}

--- a/lint/lint.go
+++ b/lint/lint.go
@@ -35,6 +35,7 @@ var checkFunctions = map[string]struct {
 	"ptr-to-ref-param":  {new: ptrToRefParamCheck},
 	"flag-deref":        {new: flagDerefCheck},
 	"switch-true":       {new: switchTrueCheck},
+	"append-combine":    {new: appendCombineCheck},
 }
 
 // RuleList returns a slice of all rules that can be used to create checkers.

--- a/lint/wrappers.go
+++ b/lint/wrappers.go
@@ -98,6 +98,37 @@ func (c baseLocalExprChecker) PerFuncInit(fn *ast.FuncDecl) bool {
 	return fn.Body != nil
 }
 
+type stmtListChecker interface {
+	CheckStmtList([]ast.Stmt)
+}
+
+type baseStmtListChecker struct {
+	ctx *context
+}
+
+// baseStmtListChecker returns a check function that visits every statement
+// list inside file. This includes block statement bodies as well as
+// implicit blocks introduced by case clauses and alike.
+func wrapBlockChecker(c stmtListChecker) func(*ast.File) {
+	return func(f *ast.File) {
+		for _, decl := range f.Decls {
+			decl, ok := decl.(*ast.FuncDecl)
+			if !ok || decl.Body == nil {
+				continue
+			}
+			ast.Inspect(decl.Body, func(x ast.Node) bool {
+				switch x := x.(type) {
+				case *ast.BlockStmt:
+					c.CheckStmtList(x.List)
+				case *ast.CaseClause:
+					c.CheckStmtList(x.Body)
+				}
+				return true
+			})
+		}
+	}
+}
+
 type stmtChecker interface {
 	PerFuncInit(*ast.FuncDecl) bool
 	CheckStmt(ast.Stmt)

--- a/lint/wrappers.go
+++ b/lint/wrappers.go
@@ -122,6 +122,8 @@ func wrapBlockChecker(c stmtListChecker) func(*ast.File) {
 					c.CheckStmtList(x.List)
 				case *ast.CaseClause:
 					c.CheckStmtList(x.Body)
+				case *ast.CommClause:
+					c.CheckStmtList(x.Body)
 				}
 				return true
 			})

